File: bindings/cpp/test/test_core.cpp
// Core KV round-trips (contract §6 "Core KV"), both backends.
#include <chrono>

#include "backends.hpp"
#include "harness.hpp"

using namespace kevy;
using namespace kevy::test;
using namespace std::chrono_literals;

KEVY_TEST(core_set_get_del_exists) {
  both_backends(t, [&](Client& c, bool) {
    c.set("k", "v");
    auto v = c.get("k");
    CHECK(v.has_value());
    CHECK_EQ(*v, std::string("v"));
    CHECK(!c.get("missing").has_value());
    CHECK_EQ(c.exists(keys({"k", "k", "missing"})), int64_t(2));  // repeated counts each
    CHECK_EQ(c.del(keys({"k"})), int64_t(1));
    CHECK(!c.get("k").has_value());
  });
}

KEVY_TEST(core_incr) {
  both_backends(t, [&](Client& c, bool) {
    CHECK_EQ(c.incr("n"), int64_t(1));
    CHECK_EQ(c.incr("n"), int64_t(2));
    CHECK_EQ(c.incr_by("n", 5), int64_t(7));
    CHECK_EQ(c.incr_by("n", -3), int64_t(4));
  });
}

KEVY_TEST(core_expire_persist_ttl) {
  both_backends(t, [&](Client& c, bool) {
    CHECK_EQ(c.ttl_ms("k"), int64_t(-2));  // no key
    c.set("k", "v");
    CHECK_EQ(c.ttl_ms("k"), int64_t(-1));  // no TTL
    CHECK(c.expire("k", 10000ms));
    CHECK(c.ttl_ms("k") > 0);
    CHECK(c.persist("k"));
    CHECK_EQ(c.ttl_ms("k"), int64_t(-1));
  });
}

KEVY_TEST(core_set_with_ttl_atomic) {
  both_backends(t, [&](Client& c, bool) {
    c.set_with_ttl("k", "v", 10000ms);
    CHECK(c.get("k").has_value());
    CHECK(c.ttl_ms("k") > 0);
  });
}

KEVY_TEST(core_type_and_dbsize_flush) {
  both_backends(t, [&](Client& c, bool) {
    c.set("s", "v");
    c.lpush("l", keys({"a"}));
    CHECK_EQ(c.type_of("s"), std::string("string"));
    CHECK_EQ(c.type_of("l"), std::string("list"));
    CHECK_EQ(c.type_of("nope"), std::string("none"));
    CHECK(c.dbsize() >= 2);
    c.flushall();
    CHECK_EQ(c.dbsize(), int64_t(0));
  });
}

KEVY_TEST(core_mget_mset) {
  both_backends(t, [&](Client& c, bool) {
    c.mset({{"a", "1"}, {"b", "2"}});
    auto got = c.mget(keys({"a", "missing", "b"}));
    CHECK_EQ(got.size(), size_t(3));
    CHECK(got[0].has_value() && *got[0] == "1");
    CHECK(!got[1].has_value());  // null preserved, not empty string
    CHECK(got[2].has_value() && *got[2] == "2");
  });
}

File: bindings/python/tests/test_core.py
"""Core KV round-trips (contract §3.1, §6) on both backends."""

import time

import kevy


def test_set_get_del_exists(backend):
    c = backend.connect()
    c.set("k", "v")
    assert c.get("k") == b"v"
    assert c.get(b"missing") is None
    assert c.exists("k", "k", "missing") == 2  # repeated key counts each time
    assert c.delete("k") == 1
    assert c.get("k") is None
    c.close()


def test_incr_family(backend):
    c = backend.connect()
    assert c.incr("n") == 1
    assert c.incr("n") == 2
    assert c.incr_by("n", 10) == 12
    assert c.incr_by("n", -5) == 7
    c.close()


def test_binary_safe_values(backend):
    c = backend.connect()
    key = bytes([0, 1, 2, 255])
    val = b"a\r\nb\x00c"
    c.set(key, val)
    assert c.get(key) == val
    c.close()


def test_expire_persist_ttl(backend):
    c = backend.connect()
    c.set("k", "v")
    assert c.ttl_ms("k") == -1  # no TTL
    assert c.ttl_ms("nope") == -2  # no key
    assert c.expire("k", 100.0) is True
    assert c.ttl_ms("k") > 0
    assert c.persist("k") is True
    assert c.ttl_ms("k") == -1
    c.close()


def test_set_with_ttl_atomic(backend):
    c = backend.connect()
    c.set_with_ttl("k", "v", 100.0)
    assert c.get("k") == b"v"
    assert c.ttl_ms("k") > 0
    c.close()


def test_type_of_and_dbsize_flushall(backend):
    c = backend.connect()
    c.set("s", "v")
    c.rpush("l", "a")
    c.hset("h", "f", "v")
    c.sadd("st", "m")
    c.zadd("z", (1.0, "m"))
    assert c.type_of("s") == "string"
    assert c.type_of("l") == "list"
    assert c.type_of("h") == "hash"
    assert c.type_of("st") == "set"
    assert c.type_of("z") == "zset"
    assert c.type_of("nope") == "none"
    assert c.dbsize() >= 5
    c.flushall()
    assert c.dbsize() == 0
    c.close()


def test_mget_mset(backend):
    c = backend.connect()
    c.mset("a", "1", "b", "2")
    assert c.mget("a", "missing", "b") == [b"1", None, b"2"]
    c.close()


def test_mset_odd_arity_rejected(backend):
    c = backend.connect()
    try:
        c.mset("a", "1", "b")
        assert False, "expected InvalidInputError"
    except kevy.InvalidInputError:
        pass
    c.close()


def test_ping(backend):
    c = backend.connect()
    c.ping()  # no error
    c.close()
